File: internal/template/template_test.go
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

package template_test

import (
	"context"
	"net/http"
	"testing"

	"github.com/steadybit/cli/internal/platformtest"
	"github.com/steadybit/cli/internal/template"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const id = "d7e65100-1d20-4980-be87-c351704910b8"

func TestListSendsTheFilters(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/templates", platformtest.Reply{JSON: map[string]any{"templates": []any{map[string]any{"id": id, "templateTitle": "Shop survives"}}}})

	out, err := platformtest.Stdout(t, func() error {
		return template.List(context.Background(), p.Client, template.ListOptions{Search: []string{"shop"}, Tags: []string{"k8s", "db"}})
	})

	require.NoError(t, err)
	assert.Contains(t, out, "Shop survives")
	q := p.Requests("GET /api/experiments/templates")[0].Query
	assert.Equal(t, []string{"shop"}, q["freeTextPhrases"])
	assert.Equal(t, []string{"k8s", "db"}, q["tag"])
}

func TestGetPlaceholdersAsAFileToFillIn(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/templates/"+id, platformtest.Reply{JSON: map[string]any{"id": id, "placeholders": []any{map[string]any{"key": "CLUSTER"}, map[string]any{"key": "NAMESPACE"}}}})

	out, err := platformtest.Stdout(t, func() error {
		return template.Get(context.Background(), p.Client, template.GetOptions{ID: id, Placeholders: true})
	})

	require.NoError(t, err)
	assert.Equal(t, "CLUSTER: ''\nNAMESPACE: ''\n\n", out)
}

func TestGetReportsAMissingTemplate(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/templates/"+id, platformtest.Reply{Status: http.StatusNotFound})

	assert.EqualError(t, template.Get(context.Background(), p.Client, template.GetOptions{ID: id}), "Experiment template "+id+" not found.")
}

File: src/template/template.test.ts
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

import { describe, expect, it, vi } from 'vitest';
import { respondTo } from '../mocks/recorder.ts';
import { getTemplate, listTemplates } from './commands.ts';

const ID = 'd7e65100-1d20-4980-be87-c351704910b8';

describe('template', () => {
  it('lists templates matching the filters', async () => {
    const requests = respondTo('get', '/api/experiments/templates', () => ({
      json: { templates: [{ id: ID, templateTitle: 'Shop survives a database outage' }] },
    }));
    const logSpy = vi.spyOn(console, 'log');

    await listTemplates({ search: ['shop'], tag: ['k8s', 'db'] });

    expect(requests[0].url.searchParams.getAll('freeTextPhrases')).toEqual(['shop']);
    expect(requests[0].url.searchParams.getAll('tag')).toEqual(['k8s', 'db']);
    expect(logSpy.mock.calls.flat().join('\n')).toContain('Shop survives a database outage');
  });

  it('prints the placeholders of a template as a file to fill in', async () => {
    respondTo('get', `/api/experiments/templates/${ID}`, () => ({
      json: {
        id: ID,
        templateTitle: 't',
        placeholders: [
          { key: 'CLUSTER', name: 'Cluster', description: '' },
          { key: 'NAMESPACE', name: 'Namespace', description: '' },
        ],
      },
    }));
    const logSpy = vi.spyOn(console, 'log');

    await getTemplate({ id: ID, placeholders: true });

    expect(logSpy).toHaveBeenCalledWith("CLUSTER: ''\nNAMESPACE: ''\n");
  });

  it('reports a template that does not exist', async () => {
    respondTo('get', `/api/experiments/templates/${ID}`, () => ({ status: 404 }));

    await expect(getTemplate({ id: ID })).rejects.toThrow(`Experiment template ${ID} not found.`);
  });
});
